language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -20,44 +24,63 @@ import           Language.PlutusCore.Lexer.Type hiding (name)
 import           Language.PlutusCore.Name
 import           Language.PlutusCore.Normalize
 import           Language.PlutusCore.Quote
-import           Language.PlutusCore.Renamer    (annotateType, rename)
+import           Language.PlutusCore.Renamer    (rename)
 import           Language.PlutusCore.Type
 import           PlutusPrelude
 
+import           Control.Lens                   (coerced)
+import           Control.Lens.TH                (makeLenses)
 import           Control.Monad.Error.Lens
 import           Control.Monad.Except
 import           Control.Monad.Reader
+import qualified Data.IntMap                    as IM
 import           Data.Map                       (Map)
 import qualified Data.Map                       as Map
 
 {- Note [Costs]
-Typechecking costs are relatively simple: it costs 1 gas to perform
-a reduction. Substitution does not in general cost anything.
+Typechecking costs are relatively simple: it costs 1 gas to perform a reduction.
+Substitution does not in general cost anything.
 
 Costs are reset every time we enter 'NormalizeTypeT'.
 
-In unlimited mode, gas is not tracked and we do not fail even on large numbers
-of reductions.
+In unlimited mode, gas is not tracked and we do not fail even on large numbers of reductions.
 -}
 
 -- | Mapping from 'DynamicBuiltinName's to their 'Type's.
 newtype DynamicBuiltinNameTypes = DynamicBuiltinNameTypes
-    { unDynamicBuiltinNameTypes :: Map DynamicBuiltinName (Quote (Type TyName ()))
+    { unDynamicBuiltinNameTypes :: Map DynamicBuiltinName (NormalizedType TyName ())
     } deriving (Semigroup, Monoid)
 
+-- TODO: define @UniqueMap@ already.
+
+newtype TyVarKinds = TyVarKinds
+    { unTyVarKinds :: IM.IntMap (Kind ())
+    } deriving (Semigroup, Monoid)
+
+newtype VarTypes = VarTypes
+    { unVarTypes :: IM.IntMap (NormalizedType TyName ())
+    } deriving (Semigroup, Monoid)  -- This is just to be able to say 'mempty'.
+                                    -- TODO: Remove this once this is not visible to the user.
+
+-- TODO: split 'TypeConfig' to 'TypeEnv' and 'TypeConfig'.
```

Comment:
:+1:

Reply:
I'll do a redesign of the type checker API. Leaking those environments to the user is no good, but leaking the annotation machinery to the user was no good as well, so I'm not making things worse here.